Sources/BraintreePayPal/BTPayPalClient.swift

```diff
@@ -0,0 +1,415 @@
+import Foundation
+import AuthenticationServices
+
+#if canImport(BraintreeCore)
+import BraintreeCore
+#endif
+
+#if canImport(BraintreeDataCollector)
+import BraintreeDataCollector
+#endif
+
+@objcMembers public class BTPayPalClient: NSObject, ASWebAuthenticationPresentationContextProviding {
+    
+    // MARK: - Internal Properties
+
+    ///Exposed for testing to get the instance of BTAPIClient
+    var apiClient: BTAPIClient
+    
+    /// Exposed for testing the approvalURL construction
+    var approvalURL: URL? = nil
+
+    /// Exposed for testing the clientMetadataID associated with this request
+    var clientMetadataID: String? = nil
+    
+    /// Exposed for testing the intent associated with this request
+    var payPalRequest: BTPayPalRequest? = nil
+
+    /// Exposed for testing, the ASWebAuthenticationSession instance used for the PayPal flow
+    var authenticationSession: ASWebAuthenticationSession? = nil
+    
+    /// Exposed for testing, for determining if ASWebAuthenticationSession was started
+    var isAuthenticationSessionStarted: Bool = false
+    
+    // MARK: - Private Properties
+
+    private var returnedToAppAfterPermissionAlert: Bool = false
+
+    // MARK: - Initializer
+
+    /// Initialize a new PayPal client instance.
+    /// - Parameter apiClient: The API Client
+    @objc(initWithAPIClient:)
+    public init(apiClient: BTAPIClient) {
+        self.apiClient = apiClient
+        super.init()
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(applicationDidBecomeActive),
+            name: UIApplication.didBecomeActiveNotification,
+            object: nil
+        )
+    }
+    
+    // MARK: - Public Methods
+    
+    /// Tokenize a PayPal account for vault or checkout.
+    ///
+    /// - Note: You can use this as the final step in your order/checkout flow. If you want, you may create a transaction from your
+    /// server when this method completes without any additional user interaction.
+    ///
+    /// On success, you will receive an instance of `BTPayPalAccountNonce`; on failure or user cancelation you will receive an error.
+    /// If the user cancels out of the flow, the error code will be `.canceled`.
+    ///
+    /// - Parameters:
+    ///   - request: Either a `BTPayPalCheckoutRequest` or a `BTPayPalVaultRequest`
+    ///   - completion: This completion will be invoked exactly once when tokenization is complete or an error occurs.
+    @objc(tokenizePayPalAccountWithPayPalRequest:completion:)
+    public func tokenizePayPalAccount(
+        with request: BTPayPalRequest & BTPayPalRequestable,
```

Comment:
Here we are requiring a request of type `BTPayPalRequest & BTPayPalRequestable` to enforce the usage of `BTPayPalCheckoutRequest` or `BTPayPalVaultRequest` and not the abstract `BTPayPalRequest`. Objective-C was more easily able to fudge the typing and [completed with an error](https://github.com/braintree/braintree_ios/blob/c449e0b81c78fcae47bb02cfb8f987aed2a19996/Sources/BraintreePayPal/BTPayPalClient.m#L107-L113) if the types did not match.

Reply:
It might be cleaner/easier for merchants to understand if we instead had 2 separate public method signatures. Something like: 
* `tokenizePayPalAccount(with vaultRequest: BTPayPalVaultRequest)`
* `tokenizePayPalAccount(with checkoutRequest: BTPayPalCheckoutRequest)`

(or whatever want to name them)

This way we can remove `BTPayPalRequestable`, which it seems just exists to enforce the type, rather than be usable to the merchant.

What do you think?

Reply:
I like this idea! Though if we want to share the code between the two we'll still need the protocol, but we can hide it from the merchant by making it private. I would also like to find a way to make `BTPayPalRequest` itself a protocol so we don't need both `BTPayPalRequest` and `BTPayPalRequestable`.

Reply:
Matt and I messed around with getting an all protocol approach for `BTPayPalRequest` to work under the current limitations of how the SDK is structured. As we move to codeable, we may have opportunities to move to more protocol oriented code as opposed to using subclassing.

Problems with protocol BTPayPalRequest (example in branch btpaypalrequest-protocol):
- Repetitive boilerplate for BTPayPalCheckoutRequest and Vault request as we cannot expose default implementations from a protocol extension in Objective-C as we can with all Swift - this limitation is a good callout if we ever need to allow Obj-C exposure on Northstar. Error with objc default protocol implementations: `Non-‘@objc’ property ‘riskCorrelationId’ does not satisfy requirement of ‘@objc’ protocol ‘BTPayPalRequest’`
- Big unwieldy initializer based on the above, we need the entirety of the init in each implementation - this makes it confusing to discern what is part of the protocol vs part of the subclass
- Cannot invoke static properties on a protocol - error `Static member 'callbackURLHostAndPath' cannot be used on protocol metatype '(any BTPayPalRequest).Type'`

Based on the above, we _did_ decide to remove the protocol `BTPayPalRequestable` all together to reduce concerns. This is reflected in commit: https://github.com/braintree/braintree_ios/pull/914/commits/a447ec02ee5f52940f2ab9c53fdd3dbf0f645c31. We decided to go with an inheritance approach because it is cleaner (less duplication of the `BTPayPalRequest` properties in the subclasses). The main downside being that we had to do some downcasting when constructing the dictionaries which we have added comments to as the alternative was to have an function that returns and empty dictionary in `BTPayPalRequest` that we would then override in the subclasses.